webserver/webserver/src/test/java/io/helidon/webserver/TransferEncodingTest.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.webserver;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import io.helidon.common.http.Http;
+import io.helidon.webserver.utils.SocketHttpClient;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.collection.IsMapContaining.hasEntry;
+
+/**
+ * Tests transfer encoding and optimizations.
+ */
+public class TransferEncodingTest {
+
+    private static final Logger LOGGER = Logger.getLogger(TransferEncodingTest.class.getName());
+
+    private static WebServer webServer;
+
+    /**
+     * Start the Web Server
+     *
+     * @param port the port on which to start the server; if less than 1,
+     * the port is dynamically selected
+     * @throws Exception in case of an error
+     */
+    private static void startServer(int port) throws Exception {
+        webServer = WebServer.create(
```

Comment:
I know this looks inspired from PlainTest, but @tomas-langer is advocating for not starting actual servers in unit tests in favor of mocking.

IIRC the idea is to have tests that do start real servers to be under /tests/functional or /tests/integration.
I don't see much under /tests at the moment, so this would mean more work to get this going.

Perhaps you can get a pass for now until we setup the integration tests properly.
Thoughts ?

Reply:
I struggle with this all the time. Recently, for another issue, I created a new functional test called `context-propagation`, but this isn't a very efficient way of writing tests because, as of now, every functional test is there for essentially _one_ purpose. If we had something more generic (like webserver) that we could easily extend, it would be a much more efficient way of writing tests.

Reply:
I agree, we need to setup a module under tests for the webserver integration / functional tests.